presto-main/src/main/java/com/facebook/presto/operator/aggregation/ReduceAggregationFunction.java

```diff
@@ -199,94 +145,52 @@ else if ((stateType.getJavaType() == Block.class)) {
         }
-        state.setLong((long) inputFunction.apply(state.getLong(), value));
-    }
-
-    public static void input(NullableDoubleState state, Object value, double initialStateValue, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
-    {
-        if (state.isNull()) {
-            state.setNull(false);
-            state.setDouble(initialStateValue);
+        try {
+            state.setValue(inputFunction.apply(state.getValue(), value));
         }
-        state.setDouble((double) inputFunction.apply(state.getDouble(), value));
-    }
-
-    public static void input(NullableBooleanState state, Object value, boolean initialStateValue, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
-    {
-        if (state.isNull()) {
-            state.setNull(false);
-            state.setBoolean(initialStateValue);
+        catch (NullPointerException npe) {
+            state.setValue(null);
         }
-        state.setBoolean((boolean) inputFunction.apply(state.getBoolean(), value));
     }
 
-    public static void input(SliceState state, Object value, Slice initialStateValue, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
+    public static void combine(ReduceAggregationState state, ReduceAggregationState otherState, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
     {
-        if (state.getSlice() == null) {
-            state.setSlice(initialStateValue);
+        if (state.getValue() == null) {
+            state.setValue(otherState.getValue());
+            return;
         }
-        state.setSlice((Slice) inputFunction.apply(state.getSlice(), value));
-    }
-
-    public static void input(BlockState state, Object value, Block initialStateValue, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
-    {
-        if (state.getBlock() == null) {
-            state.setBlock(initialStateValue);
+        try {
+            state.setValue(combineFunction.apply(state.getValue(), otherState.getValue()));
         }
-        state.setBlock((Block) inputFunction.apply(state.getBlock(), value));
-    }
-
-    public static void combine(NullableLongState state, NullableLongState otherState, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
-    {
-        if (state.isNull()) {
-            state.setNull(false);
-            state.setLong(otherState.getLong());
-            return;
+        catch (NullPointerException npe) {
+            state.setValue(null);
         }
-        state.setLong((long) combineFunction.apply(state.getLong(), otherState.getLong()));
     }
 
-    public static void combine(NullableDoubleState state, NullableDoubleState otherState, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
+    public static void write(Type type, ReduceAggregationState state, BlockBuilder blockBuilder)
```

Comment:
Shall we inline this method to `ReduceAggregationStateSerializer` and also use `writeNativeValue` (which is similar to `readNativeValue` )?

Reply:
Doesn't bindTo eleminate the redundant if's based on Type?

Reply:
Not for the function. But for the serializer I think. This `public static void write` is only used by the serializer.

Reply:
Done